// mendixmodelsdk/src/sdk/internal/properties/PrimitiveTypeEnum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrimitiveTypeEnum = void 0;
/**
 * All primitive types.
 * Note: must match exactly with MxPrimitiveTypeEnum!
 *
 * @ignore
 */
var PrimitiveTypeEnum;
(function (PrimitiveTypeEnum) {
    PrimitiveTypeEnum[PrimitiveTypeEnum["Integer"] = 0] = "Integer";
    PrimitiveTypeEnum[PrimitiveTypeEnum["String"] = 1] = "String";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Boolean"] = 2] = "Boolean";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Double"] = 3] = "Double";
    PrimitiveTypeEnum[PrimitiveTypeEnum["DateTime"] = 4] = "DateTime";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Guid"] = 5] = "Guid";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Point"] = 6] = "Point";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Size"] = 7] = "Size";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Color"] = 8] = "Color";
    PrimitiveTypeEnum[PrimitiveTypeEnum["Blob"] = 9] = "Blob";
})(PrimitiveTypeEnum = exports.PrimitiveTypeEnum || (exports.PrimitiveTypeEnum = {}));
